src/Tables.jl

```diff
@@ -31,110 +31,178 @@ function __init__()
+Obviously every table type is different, but via a combination of `Tables.rows` and `Tables.columns` each table type should be able to construct an instance of itself.
 """
 abstract type Table end
 
-abstract type AccessStyle end
-struct RowAccess <: AccessStyle end
-struct ColumnAccess <: AccessStyle end
-
-"Tables.schema(s) => NamedTuple{names, types}"
-function schema end
+# default definitions
+rowaccess(x) = false
+columnaccess(x) = false
 
 function istable(::Type{T}) where {T}
-    hasmethod(Tables.AccessStyle, Tuple{T}) &&
-    hasmethod(Tables.schema, Tuple{T}) &&
-    (Tables.AccessStyle(T) === Tables.RowAccess() ? hasmethod(Tables.rows, Tuple{T}) :
-    Tables.AccessStyle(T) === Tables.ColumnAccess() ? hasmethod(Tables.columns, Tuple{T}) : false)
+    if hasmethod(Tables.rowaccess, Tuple{T})
+        return hasmethod(Tables.rows, Tuple{T})
+    elseif hasmethod(Tables.columnaccess, Tuple{T})
+        return hasmethod(Tables.columns, Tuple{T})
+    end
+    return false
+end
+
+# Schema implementation
+"""
+    Tables.Schema(names, types)
+
+Create a `Tables.Schema` object that holds the column names and types for a tabular data object.
+`Tables.Schema` is dual-purposed: provide an easy interface for users to query these properties,
+as well as provide a convenient "structural" type for code generation. 
+
+To access the names, one can simply call `sch.names` to return the tuple of Symbols.
+To access column types, one can similarly call `sch.types`, which will return either a tuple of types (like `(Int64, Float64, String)`)
+or `nothing` if the column types are not computable or otherwise unknown.
+
+The actual type definition is
+```julia
+struct Schema{names, types} end
+```
+Where `names` is a tuple of Symbols, and `types` is a tuple _type_ of types (like `Tuple{Int64, Float64, String}`).
+Encoding the names & types as type parameters allows convenient use of the type in generated functions
+and other optimization use-cases.
+"""
+struct Schema{names, types} end
+Schema(names::Tuple{Vararg{Symbol}}, types::Type{T}) where {T <: Tuple} = Schema{names, T}()
+Schema(::Type{NamedTuple{names, types}}) where {names, types} = Schema{names, types}()
+Schema(names::Tuple{Vararg{Symbol}}, ::Nothing) = Schema{names, nothing}()
+Schema(names, ::Nothing) = Schema{Tuple(map(Symbol, names)), nothing}()
+Schema(names, types) = Schema{Tuple(map(Symbol, names)), Tuple{types...}}()
+
+function Base.getproperty(sch::Schema{names, types}, field::Symbol) where {names, types}
+    if field === :names
+        return names
+    elseif field === :types
+        return types === nothing ? nothing : Tuple(fieldtype(types, i) for i = 1:fieldcount(types))
+    else
+        throw(ArgumentError("unsupported property for Tables.Schema"))
+    end
 end
 
+function schema(x::T) where {T}
```

Comment:
Are fallbacks like this really useful? I can see a few reasons not to provide such a function:
- some types may not allow calling `first(x)` more than once, making this fallback incorrect
- it's quite trivial to implement for a specific type
- better get a `MethodError` than an empty schema for types that don't implement the interface